### PR TITLE
VSphere Lifecycle E2E

### PR DESCRIPTION
**Pull request by tatlat**
*Issue #, if available:*

*Description of changes:*
Includes tests for scaling + adding/removing worker node groups and upgrading k8s + cilium + disabling csi.

*Testing (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by cxbrowne1207 — COMMENTED** [review]
/lgtm

**Review by cxbrowne1207 — DISMISSED** [review]

**Review by cxbrowne1207 — COMMENTED** [review]
I was thinking, since we're on this test can you work on adding validations for CSI and CiliumPolicy to this? We can talk about how to do that

**Comment by cxbrowne1207** [comment]
That "approved" above was a mistake, my bad

**Comment by tatlat** [comment]
> I was thinking, since we're on this test can you work on adding validations for CSI and CiliumPolicy to this (related task is on the board)?

Yes, good idea. I can do that

File: test/framework/clustervalidator.go [hunk 1]
```diff
@@ -246,3 +247,53 @@ func getWorkerNodeMachineSets(ctx context.Context, vc ClusterValidatorConfig, w
...
+	if clusterCilium != yamlCilium {
+		return fmt.Errorf("cilium config does not match. %s and %s", clusterCilium, yamlCilium)
+	}
+
+	return nil
+}
+
+func (c *ClusterValidator) withVSphereValidations() {
+	c.WithValidation(validateCSI, 5*time.Second, 30)
+}
```
Comment (cxbrowne1207) [inline test/framework/clustervalidator.go:276]: See above comment

File: test/framework/clustervalidator.go [hunk 2]
```diff
@@ -246,3 +254,26 @@ func getWorkerNodeMachineSets(ctx context.Context, vc ClusterValidatorConfig, w
...
+	err := clusterClient.Get(ctx, key, cluster)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve cluster %s", err)
+	}
+
+	clusterCilium := cluster.Spec.ClusterNetwork.CNIConfig.Cilium.PolicyEnforcementMode
+	yamlCilium := yaml.Spec.ClusterNetwork.CNIConfig.Cilium.PolicyEnforcementMode
+	if clusterCilium != yamlCilium {
+		return fmt.Errorf("cilium config does not match. %s and %s", clusterCilium, yamlCilium)
+	}
```
Comment (cxbrowne1207) [inline test/framework/clustervalidator.go:276]: I would compare the policy in the `ClusterSpec.Cluster` to the policy referenced in `cilium-config` map instead

File: test/e2e/workload_api_test.go [hunk 3]
```diff
@@ -169,17 +169,68 @@ func TestVSphereUpgradeScaleWorkersUbuntuAPI(t *testing.T) {
...
+		wc.ValidateClusterState()
+		wc.UpdateClusterConfig(
+			api.ClusterToConfigFiller(
+				api.WithCiliumPolicyEnforcementMode(v1alpha1.CiliumPolicyModeAlways),
+			),
+			api.VSphereToConfigFiller(api.WithDisableCSI(true)),
+			vsphere.WithUbuntu124(),
+		)
+		wc.ApplyClusterManifest()
+		wc.DeleteWorkloadVsphereCSI()
```
Comment (cxbrowne1207) [inline test/e2e/workload_api_test.go:229]: Do we need to do this manually? I was under the impression that by disabling the CSI the cluster would reconcile and remove these objects automatically.

**Review by cxbrowne1207 — COMMENTED** [review]
Looking good! I left a question

File: test/e2e/workload_api_test.go (hunk 3 above)
Comment (tatlat) [reply test/e2e/workload_api_test.go:229]: Yes, I believe it has to be done manually.

**Review by cxbrowne1207 — APPROVED** [review]
/lgtm

**Pull request merged**